File: sdk/nodejs/billing_sdk.js
'use strict';
/**
 * AtonixDev Billing & Usage — Node.js Event Emission SDK
 *
 * Usage:
 *   const { BillingSDK } = require('./billing_sdk');
 *
 *   const sdk = new BillingSDK({
 *     endpoint:  'https://atonixdev.org/api/billing/events/ingest/',
 *     apiKey:    process.env.BILLING_API_KEY,
 *     service:   'compute',
 *   });
 *
 *   await sdk.emit({
 *     eventType:      'vm.started',
 *     organizationId: 'org-uuid-here',
 *     units:           1.5,
 *     unitType:        'vm_hour',
 *     metadata:        { region: 'us-east-1' },
 *   });
 *
 *   await sdk.close();
 */

const crypto  = require('crypto');
const https   = require('https');
const http    = require('http');
const { URL } = require('url');

// ─────────────────────────────────────────────────────────────────────────────
// Schema constants
// ─────────────────────────────────────────────────────────────────────────────

const REQUIRED_FIELDS = ['eventType', 'organizationId', 'units', 'unitType'];

const VALID_UNIT_TYPES = new Set([
  'vm_hour', 'gb', 'email', 'domain', 'pipeline_run',
  'gb_transfer', 'agent_hour', 'api_call', 'key',
]);

const VALID_SERVICES = new Set([
  'compute', 'storage', 'email', 'domain', 'pipeline',
  'networking', 'monitoring', 'auth', 'secrets',
]);

// ─────────────────────────────────────────────────────────────────────────────
// Validation
// ─────────────────────────────────────────────────────────────────────────────

function validateEvent(event, service) {
  for (const field of REQUIRED_FIELDS) {
    if (event[field] === undefined || event[field] === null) {
      throw new Error(`Missing required field: ${field}`);
    }
  }
  const units = Number(event.units);
  if (isNaN(units) || units < 0) {
    throw new Error('units must be a non-negative number');
  }
  if (!VALID_UNIT_TYPES.has(event.unitType)) {
    throw new Error(`Invalid unitType "${event.unitType}". Valid: ${[...VALID_UNIT_TYPES].join(', ')}`);
  }
  if (!VALID_SERVICES.has(service)) {
    throw new Error(`Invalid service "${service}". Valid: ${[...VALID_SERVICES].join(', ')}`);
  }
  if (!String(event.organizationId || '').trim()) {
    throw new Error('organizationId must be a non-empty string');
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// HTTP transport (zero dependencies)
// ─────────────────────────────────────────────────────────────────────────────

function httpPost(endpoint, apiKey, body, timeoutMs = 10000) {
  return new Promise((resolve, reject) => {
    const parsed  = new URL(endpoint);
    const isHttps = parsed.protocol === 'https:';
    const driver  = isHttps ? https : http;
    const data    = Buffer.from(JSON.stringify(body), 'utf8');

    const req = driver.request(
      {
        hostname: parsed.hostname,
        port:     parsed.port || (isHttps ? 443 : 80),
        path:     parsed.pathname + (parsed.search || ''),
        method:   'POST',
        headers:  {
          'Content-Type':   'application/json',
          'Content-Length': data.length,
          'Authorization':  `Bearer ${apiKey}`,
          'User-Agent':     'atonixdev-billing-sdk/nodejs/1.0',
        },
        timeout: timeoutMs,
      },
      (res) => {
        let raw = '';
        res.on('data', (chunk) => { raw += chunk; });
        res.on('end', () => {
          if (res.statusCode >= 400) {
            const err = new Error(`HTTP ${res.statusCode}: ${raw}`);
            err.statusCode = res.statusCode;
            return reject(err);
          }
          try { resolve(JSON.parse(raw)); }
          catch (_) { resolve(raw); }
        });
      },
    );

    req.on('timeout', () => { req.destroy(new Error('Request timed out')); });
    req.on('error', reject);
    req.end(data);
  });
}

// ─────────────────────────────────────────────────────────────────────────────
// Retry with exponential backoff
// ─────────────────────────────────────────────────────────────────────────────

async function withRetry(fn, maxRetries = 4, baseDelay = 500) {
  let delay = baseDelay;
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (err) {
      // 4xx = client error, do not retry
      if (err.statusCode && err.statusCode >= 400 && err.statusCode < 500) throw err;
      if (attempt === maxRetries) throw err;
      await new Promise(r => setTimeout(r, delay));
      delay = Math.min(delay * 2, 30000);
    }
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Deterministic event_id from content hash
// ─────────────────────────────────────────────────────────────────────────────

function deterministicId(payload) {
  const sorted = Object.keys(payload).sort().reduce((acc, k) => {
    acc[k] = payload[k]; return acc;
  }, {});
  const hash = crypto.createHash('sha256')
    .update(JSON.stringify(sorted))
    .digest('hex')
    .slice(0, 32);
  return `${hash.slice(0,8)}-${hash.slice(8,12)}-${hash.slice(12,16)}-${hash.slice(16,20)}-${hash.slice(20)}`;
}

// ─────────────────────────────────────────────────────────────────────────────
// Payload builder — maps camelCase JS API → snake_case API contract
// ─────────────────────────────────────────────────────────────────────────────

function buildPayload(event, service) {
  const payload = {
    event_type:      event.eventType,
    organization_id: event.organizationId,
    units:           event.units,
    unit_type:       event.unitType,
    service,
    timestamp:       event.timestamp || new Date().toISOString(),
  };
  if (event.eventId)    payload.event_id   = event.eventId;
  if (event.userId)     payload.user_id    = event.userId;
  if (event.projectId)  payload.project_id = event.projectId;
  if (event.metadata)   payload.metadata   = event.metadata;

  if (!payload.event_id) {
    payload.event_id = deterministicId(payload);
  }
  return payload;
}

// ─────────────────────────────────────────────────────────────────────────────
// Async batch worker
// ─────────────────────────────────────────────────────────────────────────────

class AsyncWorker {
  constructor({ transport, apiKey, batchSize = 50, flushInterval = 2000 }) {
    this._transport     = transport;
    this._apiKey        = apiKey;
    this._batchSize     = batchSize;
    this._flushInterval = flushInterval;
    this._queue         = [];
    this._pending       = 0;
    this._timer         = null;
    this._stopped       = false;
    this._drainCallbacks = [];
    this._schedule();
  }

  enqueue(payload) {
    this._queue.push(payload);
    this._pending++;
    if (this._queue.length >= this._batchSize) this._flush();
  }

  _schedule() {
    this._timer = setTimeout(() => {
      if (this._queue.length) this._flush();
      if (!this._stopped) this._schedule();
    }, this._flushInterval);
    if (this._timer.unref) this._timer.unref();
  }

  _flush() {
    if (!this._queue.length) return;
    const batch = this._queue.splice(0, this._batchSize);
    const batchApiKey = this._apiKey;
    const batchTransport = this._transport;

    withRetry(() => httpPost(
      batchTransport.replace('/ingest/', '/ingest/batch/'),
      batchApiKey,
      { events: batch },
    )).catch(err => {
      // Log but do not crash on async failure
      process.stderr.write(`[billing_sdk] async batch failed (${batch.length} events): ${err.message}\n`);
    }).finally(() => {
      this._pending -= batch.length;
      if (this._pending <= 0) {
        this._pending = 0;
        this._drainCallbacks.splice(0).forEach(cb => cb());
      }
    });
  }

  async drain() {
    if (this._pending <= 0) return;
    await new Promise(resolve => this._drainCallbacks.push(resolve));
  }

  async stop() {
    this._stopped = true;
    clearTimeout(this._timer);
    this._flush();
    await this.drain();
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Public SDK
// ─────────────────────────────────────────────────────────────────────────────

class BillingSDK {
  /**
   * @param {object} options
   * @param {string} options.endpoint    - Ingest URL
   * @param {string} options.apiKey      - Bearer token
   * @param {string} options.service     - Service name (compute/storage/…)
   * @param {boolean} [options.asyncMode=false]
   * @param {number}  [options.batchSize=50]
   * @param {number}  [options.maxRetries=4]
   * @param {number}  [options.timeout=10000]
   */
  constructor({ endpoint, apiKey, service, asyncMode = false, batchSize = 50, maxRetries = 4, timeout = 10000 }) {
    if (!VALID_SERVICES.has(service)) {
      throw new Error(`Invalid service "${service}"`);
    }
    this._endpoint   = endpoint.replace(/\/$/, '') + '/';
    this._apiKey     = apiKey;
    this._service    = service;
    this._maxRetries = maxRetries;
    this._timeout    = timeout;
    this._worker     = asyncMode
      ? new AsyncWorker({ transport: this._endpoint, apiKey, batchSize })
      : null;
  }

  /**
   * Emit a single usage event.
   *
   * Event fields (camelCase):
   *   eventType       {string}  — e.g. 'vm.started'
   *   organizationId  {string}  — org UUID
   *   units           {number}  — quantity consumed
   *   unitType        {string}  — vm_hour / gb / …
   *   eventId?        {string}  — idempotency key (auto if omitted)
   *   userId?         {string}
   *   projectId?      {string}
   *   metadata?       {object}
   *   timestamp?      {string}  — ISO8601 (defaults to now)
   *
   * @returns {Promise<object|null>} API response in sync mode, null in async.
   */
  async emit(event) {
    validateEvent(event, this._service);
    const payload = buildPayload(event, this._service);

    if (this._worker) {
      this._worker.enqueue(payload);
      return null;
    }
    return withRetry(
      () => httpPost(this._endpoint, this._apiKey, payload, this._timeout),
      this._maxRetries,
    );
  }

  /**
   * Emit multiple events as a single batch (sync only).
   * In async mode each event is individually queued.
   *
   * @param {object[]} events
   * @returns {Promise<object[]|null>}
   */
  async emitBatch(events) {
    if (this._worker) {
      for (const ev of events) await this.emit(ev);
      return null;
    }
    const payloads = events.map(ev => {
      validateEvent(ev, this._service);
      return buildPayload(ev, this._service);
    });
    return withRetry(
      () => httpPost(
        this._endpoint.replace('/ingest/', '/ingest/batch/'),
        this._apiKey,
        { events: payloads },
        this._timeout,
      ),
      this._maxRetries,
    );
  }

  /** Block until all async-queued events have been sent. */
  async flush() {
    if (this._worker) await this._worker.drain();
  }

  /** Flush and shut down the async worker. */
  async close() {
    if (this._worker) await this._worker.stop();
  }
}

module.exports = { BillingSDK, VALID_SERVICES, VALID_UNIT_TYPES };

File: sdk/python/billing_sdk.py
"""
AtonixDev Billing & Usage — Python Event Emission SDK

Usage:
    from billing_sdk import BillingSDK

    sdk = BillingSDK(
        endpoint='https://atonixdev.org/api/billing/events/ingest/',
        api_key='<your-api-key>',
        service='compute',
    )

    sdk.emit({
        'event_type':      'vm.started',
        'organization_id': 'org-uuid-here',
        'units':           1.5,
        'unit_type':       'vm_hour',
        'metadata':        {'region': 'us-east-1'},
    })
"""

import hashlib
import json
import logging
import time
import threading
import uuid
from datetime import datetime, timezone
from queue import Empty, Queue
from typing import Any, Dict, List, Optional

try:
    import urllib.request as _urllib
    import urllib.error as _urllib_err
    _HAS_URLLIB = True
except ImportError:
    _HAS_URLLIB = False

log = logging.getLogger('atonixdev.billing_sdk')

# ─────────────────────────────────────────────────────────────────────────────
# Schema
# ─────────────────────────────────────────────────────────────────────────────

REQUIRED_FIELDS = {'event_type', 'organization_id', 'units', 'unit_type'}
VALID_UNIT_TYPES = {
    'vm_hour', 'gb', 'email', 'domain', 'pipeline_run',
    'gb_transfer', 'agent_hour', 'api_call', 'key',
}
VALID_SERVICES = {
    'compute', 'storage', 'email', 'domain', 'pipeline',
    'networking', 'monitoring', 'auth', 'secrets',
}


def _validate_event(event: Dict[str, Any], service: str) -> None:
    """Raises ValueError on schema violation."""
    missing = REQUIRED_FIELDS - event.keys()
    if missing:
        raise ValueError(f'Missing required fields: {", ".join(sorted(missing))}')

    try:
        units = float(event['units'])
        if units < 0:
            raise ValueError('units must be >= 0')
    except (TypeError, ValueError) as exc:
        raise ValueError(f'units must be numeric: {exc}') from exc

    unit_type = event['unit_type']
    if unit_type not in VALID_UNIT_TYPES:
        raise ValueError(
            f'Invalid unit_type "{unit_type}". Valid: {", ".join(sorted(VALID_UNIT_TYPES))}'
        )

    if service not in VALID_SERVICES:
        raise ValueError(
            f'Invalid service "{service}". Valid: {", ".join(sorted(VALID_SERVICES))}'
        )

    if not str(event.get('organization_id', '')).strip():
        raise ValueError('organization_id must be a non-empty string')


# ─────────────────────────────────────────────────────────────────────────────
# HTTP Transport
# ─────────────────────────────────────────────────────────────────────────────

class _Transport:
    def __init__(self, endpoint: str, api_key: str, timeout: int = 10):
        self.endpoint = endpoint.rstrip('/')
        self.api_key  = api_key
        self.timeout  = timeout

    def send(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        body    = json.dumps(payload).encode('utf-8')
        headers = {
            'Content-Type':  'application/json',
            'Authorization': f'Bearer {self.api_key}',
            'User-Agent':    'atonixdev-billing-sdk/python/1.0',
        }
        req = _urllib.Request(self.endpoint + '/', data=body, headers=headers, method='POST')
        try:
            with _urllib.urlopen(req, timeout=self.timeout) as resp:
                return json.loads(resp.read().decode('utf-8'))
        except _urllib_err.HTTPError as exc:
            body = exc.read().decode('utf-8', errors='replace')
            raise RuntimeError(f'HTTP {exc.code}: {body}') from exc

    def send_batch(self, payloads: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        body    = json.dumps({'events': payloads}).encode('utf-8')
        headers = {
            'Content-Type':  'application/json',
            'Authorization': f'Bearer {self.api_key}',
            'User-Agent':    'atonixdev-billing-sdk/python/1.0',
        }
        req = _urllib.Request(
            self.endpoint.replace('/ingest/', '/ingest/batch/') + '/',
            data=body, headers=headers, method='POST',
        )
        try:
            with _urllib.urlopen(req, timeout=self.timeout) as resp:
                return json.loads(resp.read().decode('utf-8'))
        except _urllib_err.HTTPError as exc:
            body = exc.read().decode('utf-8', errors='replace')
            raise RuntimeError(f'HTTP {exc.code}: {body}') from exc


# ─────────────────────────────────────────────────────────────────────────────
# Retry logic
# ─────────────────────────────────────────────────────────────────────────────

def _with_retry(fn, max_retries: int = 4, base_delay: float = 0.5):
    """Exponential backoff retry — does NOT retry on 4xx (client errors)."""
    delay = base_delay
    for attempt in range(max_retries + 1):
        try:
            return fn()
        except RuntimeError as exc:
            msg = str(exc)
            # 4xx = caller error, do not retry
            if msg.startswith('HTTP 4'):
                raise
            if attempt == max_retries:
                raise
            log.warning('billing_sdk retry %d/%d after error: %s', attempt + 1, max_retries, msg)
            time.sleep(delay)
            delay = min(delay * 2, 30.0)


# ─────────────────────────────────────────────────────────────────────────────
# Async batch worker
# ─────────────────────────────────────────────────────────────────────────────

class _AsyncWorker(threading.Thread):
    """Background thread that drains an in-memory queue in micro-batches."""

    def __init__(self, transport: _Transport, service: str,
                 batch_size: int = 50, flush_interval: float = 2.0):
        super().__init__(daemon=True, name='billing-sdk-worker')
        self._q              = Queue()
        self._transport      = transport
        self._service        = service
        self._batch_size     = batch_size
        self._flush_interval = flush_interval
        self._stop_event     = threading.Event()

    def enqueue(self, payload: Dict[str, Any]) -> None:
        self._q.put(payload)

    def flush(self) -> None:
        """Block until the queue is empty."""
        self._q.join()

    def stop(self) -> None:
        self._stop_event.set()
        self.flush()

    def run(self) -> None:
        while not self._stop_event.is_set():
            batch: List[Dict[str, Any]] = []
            deadline = time.monotonic() + self._flush_interval

            # Collect up to batch_size events or until flush_interval
            while len(batch) < self._batch_size:
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    break
                try:
                    item = self._q.get(timeout=min(remaining, 0.1))
                    batch.append(item)
                except Empty:
                    if time.monotonic() >= deadline:
                        break

            if batch:
                self._send_batch(batch)

        # Drain remaining on shutdown
        remaining: List[Dict[str, Any]] = []
        while not self._q.empty():
            try:
                remaining.append(self._q.get_nowait())
            except Empty:
                break
        if remaining:
            self._send_batch(remaining)

    def _send_batch(self, batch: List[Dict[str, Any]]) -> None:
        try:
            _with_retry(lambda: self._transport.send_batch(batch))
        except Exception as exc:  # noqa: BLE001
            log.error('billing_sdk async batch send failed (%d events): %s', len(batch), exc)
        finally:
            for _ in batch:
                self._q.task_done()


# ─────────────────────────────────────────────────────────────────────────────
# Public SDK
# ─────────────────────────────────────────────────────────────────────────────

class BillingSDK:
    """
    Thread-safe AtonixDev Billing SDK.

    Parameters
    ----------
    endpoint  : Full URL to the ingest endpoint
                e.g. 'https://atonixdev.org/api/billing/events/ingest/'
    api_key   : Bearer token for the billing API
    service   : Service identifier (compute/storage/email/…)
    async_mode: If True, events are queued and sent in background batches.
                If False (default), each call blocks until the event is delivered.
    batch_size: Max events per async batch (default 50)
    max_retries: Retry attempts on transient failures (default 4)
    timeout   : HTTP request timeout in seconds (default 10)
    """

    def __init__(
        self,
        endpoint:    str,
        api_key:     str,
        service:     str,
        async_mode:  bool  = False,
        batch_size:  int   = 50,
        max_retries: int   = 4,
        timeout:     int   = 10,
    ):
        if service not in VALID_SERVICES:
            raise ValueError(f'Invalid service "{service}"')

        self._service     = service
        self._max_retries = max_retries
        self._transport   = _Transport(endpoint, api_key, timeout)
        self._async       = async_mode
        self._worker: Optional[_AsyncWorker] = None

        if async_mode:
            self._worker = _AsyncWorker(self._transport, service, batch_size)
            self._worker.start()

    # ------------------------------------------------------------------
    def emit(self, event: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """
        Emit a single usage event.

        Required fields in *event*:
          event_type      str  — e.g. 'vm.started'
          organization_id str  — org UUID
          units           num  — quantity consumed
          unit_type       str  — vm_hour / gb / email / …

        Optional:
          event_id   str   — idempotency key (auto-generated if omitted)
          user_id    str   — user who triggered the event
          project_id str   — project context
          metadata   dict  — arbitrary key/value pairs
          timestamp  str   — ISO8601 (defaults to now)

        Returns the API response dict in sync mode, None in async mode.
        """
        _validate_event(event, self._service)
        payload = self._build_payload(event)

        if self._async and self._worker:
            self._worker.enqueue(payload)
            return None
        return _with_retry(
            lambda: self._transport.send(payload),
            max_retries=self._max_retries,
        )

    def emit_batch(self, events: List[Dict[str, Any]]) -> Optional[List[Dict[str, Any]]]:
        """
        Emit a list of events as a single batch request (sync).

        No-op in async mode (events are already batched by the worker).
        """
        if self._async:
            for ev in events:
                self.emit(ev)
            return None

        validated = []
        for ev in events:
            _validate_event(ev, self._service)
            validated.append(self._build_payload(ev))

        return _with_retry(
            lambda: self._transport.send_batch(validated),
            max_retries=self._max_retries,
        )

    def flush(self) -> None:
        """Block until all async-queued events have been delivered."""
        if self._worker:
            self._worker.flush()

    def close(self) -> None:
        """Flush and shut down the async worker."""
        if self._worker:
            self._worker.stop()

    # ------------------------------------------------------------------
    def _build_payload(self, event: Dict[str, Any]) -> Dict[str, Any]:
        """Attach service identity, timestamp, and idempotency key."""
        payload = dict(event)
        payload['service'] = self._service

        # Idempotency key — deterministic from content if not supplied
        if 'event_id' not in payload:
            fingerprint = json.dumps(
                {k: payload[k] for k in sorted(payload)}, sort_keys=True, default=str
            )
            payload['event_id'] = str(
                uuid.UUID(hashlib.sha256(fingerprint.encode()).hexdigest()[:32])
            )

        # Timestamp
        if 'timestamp' not in payload:
            payload['timestamp'] = datetime.now(timezone.utc).isoformat()

        return payload

    # Context-manager support
    def __enter__(self):
        return self

    def __exit__(self, *_):
        self.close()
